File: TP2/algo1.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int rechElets_TabNonTries(int tab[], int n, int x) {
    for (int i = 0; i < n; i++) {
        if (tab[i] == x)
            return 1; // trouvé
    }
    return 0; // non trouvé
}

int main() {
    FILE *f_tailles = fopen("tailles.txt", "r");
    FILE *f_res = fopen("resultats1.csv", "w");
    if (!f_tailles || !f_res) {
        printf("Erreur lors de l'ouverture d'un fichier.\n");
        return 1;
    }

    fprintf(f_res, "n,temps_meilleur,temps_pire\n");

    int n;
    srand(time(NULL));

    while (fscanf(f_tailles, "%d", &n) == 1) {
        int *tab = (int*) malloc(n * sizeof(int));
        if (!tab) continue;

        // Remplir le tableau avec des valeurs aléatoires
        for (int i = 0; i < n; i++)
            tab[i] = rand() % (n * 10);

        clock_t debut, fin;
        double temps_meilleur, temps_pire;

        // ---- Meilleur cas (élément au début)
        int x_meilleur = tab[0];
        debut = clock();
        rechElets_TabNonTries(tab, n, x_meilleur);
        fin = clock();
        temps_meilleur = (double)(fin - debut) / CLOCKS_PER_SEC;

        // ---- Pire cas (élément absent)
        int x_pire = -1; // on choisit un nombre qui ne peut pas exister
        debut = clock();
        rechElets_TabNonTries(tab, n, x_pire);
        fin = clock();
        temps_pire = (double)(fin - debut) / CLOCKS_PER_SEC;

        // Sauvegarde des résultats
        fprintf(f_res, "%d,%f,%f\n", n, temps_meilleur, temps_pire);

        free(tab);
    }

    fclose(f_tailles);
    fclose(f_res);
    printf("Mesures terminées. Résultats enregistrés dans resultats.csv\n");
    return 0;
}

File: TP2/algo1.py
import time
import csv

# --- Fonction de recherche dichotomique ---
def rechElets_Dicho(tab, x):
    debut = 0
    fin = len(tab) - 1
    while debut <= fin:
        milieu = (debut + fin) // 2
        if tab[milieu] == x:
            return True
        elif tab[milieu] < x:
            debut = milieu + 1
        else:
            fin = milieu - 1
    return False

# --- Fonction pour mesurer le temps moyen ---
def mesure_temps(func, *args, repetitions=1000):
    total = 0
    for _ in range(repetitions):
        start = time.perf_counter()
        func(*args)
        end = time.perf_counter()
        total += (end - start)
    return total / repetitions

# --- Fichiers ---
with open("tailles.txt", "r") as f_tailles, open("resultats_dicho_precis.csv", "w", newline='') as f_res:
    writer = csv.writer(f_res)
    writer.writerow(["n", "temps_meilleur", "temps_pire"])
    
    for ligne in f_tailles:
        n = int(ligne.strip())
        tab = [i * 2 for i in range(n)]  # tableau trié

        # Meilleur cas : élément du milieu
        x_meilleur = tab[len(tab) // 2]
        temps_meilleur = mesure_temps(rechElets_Dicho, tab, x_meilleur, repetitions=1000)

        # Pire cas : élément absent
        x_pire = -1
        temps_pire = mesure_temps(rechElets_Dicho, tab, x_pire, repetitions=1000)

        writer.writerow([n, temps_meilleur, temps_pire])

print("Mesures terminées. Résultats enregistrés dans resultats_dicho_precis.csv")

